blog/2022-05-13-caching-api-requests-with-long-lived-workflows.md

```diff
@@ -0,0 +1,231 @@
+  "rates": {
+    "AED": 3.6731,
+    "AFN": 87.249991,
+    "ALL": 114.227196,
+    "AMD": 473.915796,
+    ...
+  },
+  "lastupdate": "2022-05-11T00:46:04.483000+00:00"
+}
+```
+
+Because Temporal Workflows must be deterministic, you can't make an API request directly from a Workflow.
+You need to create an Activity that makes an API request, and call that Activity from your Workflow.
+Below is an [`activities.ts` file](https://github.com/vkarpov15/temporal-api-caching-example/blob/main/src/activities.ts) that makes an HTTP request to the above API endpoint, and returns the exchange rates.
+
+```ts
+import axios from 'axios';
+
+export async function getExchangeRates(): Promise<any> {
+  const res = await axios.get('https://cdn.moneyconvert.net/api/latest.json');
+  return res.data.rates;
+};
+```
+
+Next, let's write a Workflow that calls this Activity once per day.
+With Temporal Workflows, you can simply write a `while (true) {}` loop with a [JavaScript sleep](https://masteringjs.io/tutorials/fundamentals/sleep) that pauses the Workflow until the next time you need to refresh exchange rates.
+Writing the Workflow this way may seem counterintuitive, because we've all had to learn over the course of our careers that writing applications isn't this easy.
+But, with Temporal, it actually is this easy!
+
+```ts
+import {
+  defineQuery,
+  proxyActivities,
+  setHandler,
+  sleep
+} from '@temporalio/workflow';
+import type * as activities from './activities';
+
+const { getExchangeRates } = proxyActivities<typeof activities>({
+  startToCloseTimeout: '1 minute',
+});
+
+export const getExchangeRatesQuery = defineQuery<any, [string]>('getExchangeRates');
+
+export async function exchangeRatesWorkflow(): Promise<any> {
+  let rates: any = null;
+
+  // Register a query handler that allows querying for the current rates
+  setHandler(getExchangeRatesQuery, () => rates);
+
+  while (true) {
+    // Get the latest rates
+    rates = await getExchangeRates();
+
+    // Sleep until tomorrow at 12pm server time, and then get the rates again
+    const today = new Date();
+    const tomorrow = new Date(today);
+    tomorrow.setHours(12, 0, 0, 0);
+    tomorrow.setDate(tomorrow.getDate() + 1);
+    // @ts-ignore
+    await sleep(tomorrow - today);
+  }
+}
+```
+
+That's the entire Workflow for caching exchange rates.
+The full source code is available in [this GitHub repo](https://github.com/vkarpov15/temporal-api-caching-example).
+Notice there's no explicit references to a database or job queue.
+This Workflow is almost pure business logic, with a minimum of references to frameworks or services.
+
+To run this Workflow, you can run a `start-workflow.ts` script as shown below.
+This script starts a Workflow and exits, leaving the Workflow running on the Worker.
+Note that only one Workflow with a given `workflowId` can run at any given time, so the below code also ensures that only one copy of this Workflow is running at any given time.
```

Comment:
```suggestion
Note that only one Workflow with a given `workflowId` can run at any given time, so the following code also ensures that only one copy of this Workflow is running at any particular time.
```
Last edit per [given](https://docs.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/g/given) in MWSG.